### ISSUE TITLE
rootfs build errors and silently ignored errors

### ISSUE THREAD

**Issue by sdmaclea** [dotnet/dotnet-buildtools-prereqs-docker#22]
Looking at CI logs, all logs show `"arm" and "arm64" rootfs builds having errors.  However, the errors are silently ignored and the build succeeds.  

See debug notes in https://github.com/dotnet/coreclr/issues/16900

I looked at the visible console logs starting at build #5.  I see the error for all "arm" and "arm64" attempts to build a rootfs.  

The errors occur when attempting to run the initial qemu-debootstrap command.  Missing package?

@michaelsimons @janvorli

**Comment by jashook** [comment]
Transferring some of the conversation to this issue.  Note this is also blocking the cross x86/arm work @echesakovMSFT is doing.

Below is a statement from https://github.com/dotnet/coreclr/issues/16900.
```
Looks like @janvorli is right those binfmt errors are problematic.

http://neophob.com/2014/10/run-arm-binaries-in-your-docker-container-using-boot2docker/

There was also hint elsewhere that configuring the host to mount binfmt may also fix the issue
```

@sdmaclea @MichaelSimons do we need to build the rootfs under docker? It seems like a lot of current/future problems could be solved by building them on the host then copying them into the docker. I guess the largest issue is the rootfs build script requires root access.

**Comment by sdmaclea** [comment]
> do we need to build the rootfs under docker?

We could use a rootfs tar image, but I wonder if we aren't just deferring the problem.  Ultimately we need to be able to build from source.  So we need a method to repeatably build the rootfs tar image as well, which probably leads us back to docker...

**Comment by MichaelSimons** [comment]
@jashook - The reason the rootfs was built in a docker container in the first place is it is great way to establish the dependencies/configuration necessary to build the rootfs.  This is the very reason we run our builds in a container.  Without this we'd need a "specialized" build machine available to us with the appropriate configuration.  This is not a path I am interested in taking.

**Comment by shawnro** [comment]
@MichaelSimons, it sounds like we won't have the ideal long term solution for this in place anytime soon.  What's the right path forward here so we can get things unblocked?

**Comment by jashook** [comment]
Using a local build I cannot repro the failure to chroot which is output below:

```
18:06:46 I: Running command: chroot /rootfs/arm64 /debootstrap/debootstrap --second-stage
18:06:46 chroot: failed to run command '/debootstrap/debootstrap': Exec format error
18:06:46 /scripts/cross/build-rootfs.sh: line 138: install_Failed: command not found
18:06:46 chroot: failed to run command 'apt-get': No such file or directory
18:06:46 /scripts/cross/build-rootfs.sh: line 146: install_Failed: command not found
18:06:46 chroot: failed to run command 'apt-get': No such file or directory
18:06:46 /scripts/cross/build-rootfs.sh: line 150: install_Failed: command not found
18:06:46 chroot: failed to run command 'apt-get': No such file or directory
18:06:46 /scripts/cross/build-rootfs.sh: line 154: install_Failed: command not found
18:06:46 chroot: failed to run command 'symlinks': No such file or directory
18:06:46 /scripts/cross/build-rootfs.sh: line 158: install_Failed: command not found
```

Minus the binfmt error:

```
18:17:16 Setting up binfmt-support (2.1.6-1) ...
18:17:16 update-binfmts: warning: unable to open /proc/filesystems: No such file or directory
18:17:16 update-binfmts: warning: Couldn't load the binfmt_misc module.
18:17:16 update-binfmts: warning: unable to open /proc/filesystems: No such file or directory
18:17:16 update-binfmts: warning: Couldn't load the binfmt_misc module.
18:17:16 update-binfmts: warning: unable to open /proc/filesystems: No such file or directory
18:17:16 update-binfmts: warning: Couldn't load the binfmt_misc module.
```

I am able to build using docker 17.12.0-ce on ubuntu 16.04, building ubuntu 16.04 image with the arm rootfs.

Currently I am re-syncing and rebuilding for arm64 to see if the combo changes the outcome. In the meantime is there a repro machine I can use from the lab? @mmitche

**Comment by sdmaclea** [comment]
When I pull the `ubuntu-16.04-cross-arm64-d809634-20180312171656` image the `/crossrootfs` directory is empty.  Also the `/proc/sys/fs/bunfmt_misc` filesystem is empty

When I pull the `ubuntu-16.04-crossdeps-d770b8b-20180312171656` image the `/proc/sys/fs/bunfmt_misc` filesystem is empty.  While running interactive in `--privileged`, attempting to install `binfmt-support` is ignored because it is already installed.  If I uninstall and reinstall, the rootfs seems to be build correctly.

I suspect the docker host machine must have binfmt-support installed.  Perhaps the machine select in `netci.groovy` needs to be changed.

**Comment by sdmaclea** [comment]
Something like  `Utilities.setMachineAffinity(newJob, 'Ubuntu14.04', 'arm-cross-latest')` instead of 
https://github.com/dotnet/dotnet-buildtools-prereqs-docker/blob/master/netci.groovy#L19

Which is what corefx uses when running the cross docker images

**Comment by sdmaclea** [comment]
To be complete when I pull the `ubuntu-16.04-cross-d809634-20180312171656` image the `/crossrootfs` directory is empty.  It should have `x86` and `arm`

I don't plan to work on this anymore.  I do not think I can fix the issue which I suspect is a machine affinity issue as suggested above.

**Comment by sdmaclea** [comment]
The working arm image:
`microsoft/dotnet-buildtools-prereqs:ubuntu-14.04-cross-0cd4667-20170219020255` 
has a populated `/crossrootfs/arm` directory.  
It does not have a populated `/proc/sys/fs/binfmt_misc` at least out of the box.

**Comment by mmitche** [comment]
@sdmaclea We can get the bin utils on those images. I'd rather not use the arm-cross-latest image as that image has some serious issues allocating reliably.

Btw, @jashook I was thinking of replacing that arm-cross-latest image (it has ~10% allocation success rate). It was the one with the specialized tizen setup right?

**Comment by sdmaclea** [comment]
@mmitche I am not 100% certain the issues is missing packages on the docker host, but it seems like a reasonable theory.  It would explain why when @jashook and I try to reproduce on machine we have configured to run build-rootfs.sh the problem are not reproduced.

**Comment by sdmaclea** [comment]
The errors are no longer silently ignored.  Fixed by https://github.com/dotnet/core-setup/pull/3831

**Comment by jashook** [comment]
These machines have binfmt-support installed. I do not believe that is the issue.

**Comment by jashook** [comment]
The docker version is not the root cause, I have reproed the failure with docker version 17.12.1-ce

**Comment by sdmaclea** [comment]
@jashook Does the host need qemu ?

**Comment by jashook** [comment]
It is possible, have installed qemu, will rebuild

**Comment by jashook** [comment]
I went ahead and installed qemu qemu-user-static b debootstrap. With those packages chroot succeeds and the rootfs build continues.

**Comment by jashook** [comment]
Will work with @mmitche to update the base image.

**Comment by MichaelSimons** [comment]
@jashook -  In addition to CI, we would also need a build machine with this configuration.

**Comment by mmitche** [comment]
@MichaelSimons what official pool are the docker builds done in?

**Comment by jashook** [comment]
@mmitche they are in Dotnet-Build

**Comment by MichaelSimons** [comment]
I'm concerned there is another issue that will get hit once the host configuration issues are resolved.  The local volume mount used in the [build-rootfs.sh](https://github.com/dotnet/dotnet-buildtools-prereqs-docker/blob/master/src/ubuntu/build-scripts/build-rootfs.sh#L29) isn't going to work because it is being executed in a nested container scenario.  This is not supported yet fails silently. 

It would be a good idea to add a check to verify the rootfs folder is not empty at the end of this script.

This issue is easily resolved with something like [this](https://github.com/MichaelSimons/dotnet-buildtools-prereqs-docker/commit/aa249eb6a87e9fdc6eb6dc13f7fb155af8c36dbb)

**Comment by jashook** [comment]
With the latest image @mmitche added to the ci the build has successfully mounted the rootfs and will continue building correctly. See https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/23 and https://ci.dot.net/job/dotnet_dotnet-buildtools-prereqs-docker/job/master/job/ubuntu_16_prtest/18/console.

**Comment by jashook** [comment]
Build agent changes will unblock the official build, we can address the issue Michael pointed out in parallel. @mmitche can we start the official build machine update?

**Comment by mmitche** [comment]
Submitted tickets

**Comment by MichaelSimons** [comment]
@mmitche - Any update on when an official build machine will be available?

**Comment by jashook** [comment]
The build machines were updated. The last packages built by #23 

```
microsoft/dotnet-buildtools-prereqs:ubuntu-14.04-a3ae44b-20180316024003
microsoft/dotnet-buildtools-prereqs:ubuntu-14.04-debpkg-d770b8b-20180316024003
microsoft/dotnet-buildtools-prereqs:ubuntu-14.04-coredeps-d770b8b-20180316024003
microsoft/dotnet-buildtools-prereqs:ubuntu-14.04-crossdeps-d770b8b-20180316024003
microsoft/dotnet-buildtools-prereqs:ubuntu-14.04-cross-a3ae44b-20180316024003
microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-a3ae44b-20180316023254
microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-debpkg-d770b8b-20180316023254
microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-coredeps-d770b8b-20180316023254
microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-crossdeps-620ba25-20180316023254
microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-cross-arm64-a3ae44b-20180316023254
microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-cross-a3ae44b-20180316023254
```

Are missing the x86 rootfs.

**Comment by jashook** [comment]
Looks like when we add https://github.com/dotnet/dotnet-buildtools-prereqs-docker/blob/master/src/ubuntu/16.04/cross/Dockerfile#L3 we are only adding one of the rootfs tar files.

**Comment by sdmaclea** [comment]
My apparently faulty assumption was that the two arches would be concatenated by the `>>` in 
https://github.com/dotnet/dotnet-buildtools-prereqs-docker/blob/master/src/ubuntu/build-scripts/build-rootfs.sh#L65

**Comment by jashook** [comment]
Fixed by #23 and #27
